### DISCUSSION TITLE
Managing Cookies in Cypress tests

### DISCUSSION THREAD

**Discussion by williamtaggart97** [cypress-io/cypress#30413] [Questions and Help]
I'm working through setting up login for my E2E tests, but the login/callback relies on an `oauth-state` cookie that isn't there when I log in via Cypress.

Here's the login code: 
```
Cypress.Commands.add('login', (email, password) => {
  cy.visit('/')
  cy.get('input#username').type(email)
  cy.get('input#password').type(password)
  cy.contains('button[value=default]', 'Continue').click()
})
```

And here's the next route that's throwing an error (I've confirmed that the query params for `code` and `state` are coming through correctly)
```
export async function GET(req: NextRequest) {
  const url = new URL(req.url)
  const code = url.searchParams.get('code')
  const state = url.searchParams.get('state')

  const storedState = req.cookies.get('canvas-oauth-state')?.value ?? null

  if (!code || !state || state !== storedState) {
    return NextResponse.json(
      { message: 'Invalid Oauth2 state' },
      { status: 400 }
    )
  }
```

My question is: How can I make sure that the `canvas-oauth-state` cookie is successfully passed to the callback route? 

I've tried looking through the cookies using `Cypress.Cookies.debug(true)` and I've tried logging cookies at different steps in the process and it's really weird. When I am going through the log in process, I see that the debug logs remove an old cookie and set a new cookie, so I know it exists somewhere in the process. HOWEVER, whenever I try to add manual logs in my login action to see if I can manipulate that cookie, it's not there... 

I've used `cy.getCookies`, `cy.getAllCookies` and `cy.getCookie('canvas-oauth-state')` and they never return the object you see in the screenshot below. 

<img width="1512" alt="Screenshot 2024-10-16 at 4 33 28 PM" src="https://github.com/user-attachments/assets/95b7047d-46f0-4ace-b0e0-70849a9cecd8">

Any input on the discrepancies between the debug logs and `cy.getCookie` would be appreciated. I've been thinking about maybe a change in domain, but I'm lost...

**Comment by portyu9** [comment]
Hi @williamtaggart97 

Since cy.getAllCookies() also does not return canvas-oauth-state, this is not just cy.getCookie() looking at the wrong current domain.

Cypress.Cookies.debug(true) logs cookie modifications, so seeing the cookie being set there does not necessarily mean it still exists when the OAuth callback runs.

I would inspect the actual Set-Cookie that creates canvas-oauth-state and verify its:

Domain
Path
SameSite
Secure
Expires / Max-Age

Then check whether the cookie exists immediately after it is created and whether the callback request actually contains it in its Cookie header.

If it appears and later disappears, determine what response/navigation clears or expires it. If it never appears in cy.getAllCookies(), determine why the browser rejected or failed to persist the Set-Cookie. If it exists but is absent from the callback request, its domain/path/SameSite/Secure scope is the issue.

Because this is an OAuth state cookie, I would not fix the test by manually recreating it with cy.setCookie()—that would bypass the state mechanism you're trying to verify.

**Reply by portyu9** [reply]
@williamtaggart97 Please mark this answer as accepted if it helped, thank you.
